File: src-tauri/src/commands/repo.rs
//! Repository lifecycle: open, graph/history reads, clone/init, recents, and the filesystem watch.

use super::{blocking, sync, CommandError};
use crate::git::types::{
    BranchInfo, GitTransportAuthRef, HistorySearchPage, HistorySearchQuery, HistorySearchResult,
    RecentStatus, RepoGraph, RepoSummary,
};
use crate::watcher::WatcherState;
use crate::{git, shell, watcher};

/// Holds the in-flight `git clone` child so [`cancel_clone`] can terminate it
/// from another command while the clone streams progress. The inner [`Arc`] is
/// cloned out before the clone runs on the blocking pool.
#[derive(Default)]
pub struct CloneState(git::write::lifecycle::CloneSlot);

/// Initial graph window. The frontend explicitly increases this in 2,000-commit
/// pages while virtualized rows/canvas keep rendering memory bounded.
const DEFAULT_GRAPH_LIMIT: usize = 2000;

#[tauri::command]
pub fn open_repo(path: String) -> Result<RepoSummary, CommandError> {
    // `RepoOpenError` converts into `CommandError` with `kind` missingPath /
    // notARepository and the attempted `path`, so the frontend can tell a
    // moved/deleted repository apart from a real failure and offer the
    // dedicated missing-repo state with Remove / Locate / Retry (GL-108).
    sync(|| git::read::summary_classified(&path))
}

#[tauri::command]
pub async fn commit_graph(path: String, limit: Option<usize>) -> Result<RepoGraph, CommandError> {
    let limit = limit.unwrap_or(DEFAULT_GRAPH_LIMIT);
    // Large histories can spend hundreds of milliseconds in ref collection,
    // revwalk, lane layout, and serialization. Open the non-Send Repository
    // inside the worker closure so none of that blocks the webview thread.
    blocking(move || git::read::commit_graph(&path, limit).map_err(|e| e.to_string())).await
}

#[tauri::command]
pub async fn search_history(
    path: String,
    query: HistorySearchQuery,
) -> Result<HistorySearchPage, CommandError> {
    blocking(move || git::read::search_history(&path, query)).await
}

#[tauri::command]
pub async fn suggest_tree_paths(
    path: String,
    filter: String,
    limit: Option<usize>,
) -> Result<Vec<String>, CommandError> {
    // The HEAD tree walk is proportional to the repo's file count — keep it
    // off the webview thread like the other potentially expensive reads.
    blocking(move || git::read::suggest_tree_paths(&path, &filter, limit)).await
}

#[tauri::command]
pub fn list_branches(path: String) -> Result<Vec<BranchInfo>, CommandError> {
    sync(|| git::read::branches(&path).map_err(|e| e.to_string()))
}

#[tauri::command]
pub fn can_fast_forward(path: String, from: String, to: String) -> Result<bool, CommandError> {
    sync(|| git::read::can_fast_forward(&path, &from, &to).map_err(|e| e.to_string()))
}

/// The commits `base..head` would carry, newest first.
#[tauri::command]
pub async fn range_commits(
    path: String,
    base: String,
    head: String,
) -> Result<Vec<HistorySearchResult>, CommandError> {
    blocking(move || git::read::range_commits(&path, &base, &head)).await
}

/// Which of `candidates` `head` descends from, nearest first.
#[tauri::command]
pub async fn ancestor_refs(
    path: String,
    head: String,
    candidates: Vec<String>,
) -> Result<Vec<String>, CommandError> {
    blocking(move || git::read::ancestor_refs(&path, &head, &candidates)).await
}

/// The branch a new pull request from `head` should target by default.
#[tauri::command]
pub async fn default_base_branch(
    path: String,
    head: String,
) -> Result<Option<String>, CommandError> {
    blocking(move || git::read::default_base_branch(&path, &head)).await
}

/// Clone `url` into `dest`, streaming `clone-progress` events as it runs. The
/// child is parked in [`CloneState`] so [`cancel_clone`] can stop it. Returns the
/// cloned path on success; the frontend then opens it.
#[tauri::command]
pub async fn clone_repo(
    app: tauri::AppHandle,
    state: tauri::State<'_, CloneState>,
    url: String,
    dest: String,
    auth: Option<GitTransportAuthRef>,
) -> Result<String, CommandError> {
    let slot = state.0.clone();
    blocking(move || {
        let cred = git::transport_auth::credential_for_url(&url, auth.as_ref())?;
        // A dropped progress tick must never fail the clone itself.
        let progress = |p: &git::write::lifecycle::CloneProgress| {
            crate::events::emit(&app, crate::events::CLONE_PROGRESS, p.clone());
        };
        git::write::lifecycle::clone(&progress, slot, &url, &dest, &cred)
    })
    .await
}

/// Terminate an in-flight [`clone_repo`]. Instant (lock + kill), so it stays a
/// plain sync command and never queues behind the blocking pool the clone holds.
#[tauri::command]
pub fn cancel_clone(state: tauri::State<'_, CloneState>) -> Result<(), CommandError> {
    sync(|| git::write::lifecycle::cancel_clone(&state.0))
}

/// Initialize a new repository at `parent`/`name` on `branch`, optionally seeding
/// a README and a `.gitignore` template. Returns the new repo's path.
#[tauri::command]
pub async fn init_repo(
    parent: String,
    name: String,
    branch: String,
    readme: bool,
    gitignore: String,
) -> Result<String, CommandError> {
    blocking(move || git::write::lifecycle::init(&parent, &name, &branch, readme, &gitignore)).await
}

/// Initialize an already-existing, possibly non-empty directory as a git
/// repository in place — the missing-repo screen's "Initialize as git repo"
/// recovery action for a folder that lost its `.git` (GL-153).
#[tauri::command]
pub async fn init_repo_in_place(path: String) -> Result<String, CommandError> {
    blocking(move || git::write::lifecycle::init_in_place(&path)).await
}

/// Presence + current branch for each recent repo path, so the onboarding list
/// can flag missing paths and show branches. Touches the filesystem per path, so
/// it runs off the main thread.
#[tauri::command]
pub async fn recents_status(paths: Vec<String>) -> Result<Vec<RecentStatus>, CommandError> {
    blocking(move || Ok::<_, CommandError>(git::read::recents_status(&paths))).await
}

/// Reveal `path` in the OS file manager (Finder/Explorer). Spawns and returns
/// immediately, so a plain sync command is fine.
#[tauri::command]
pub fn reveal_path(path: String) -> Result<(), CommandError> {
    sync(|| shell::reveal(&path))
}

/// Start (or replace) the filesystem watch for `path`, emitting path-tagged
/// `repo-changed` events when its worktree or git state changes. Each open
/// tab keeps its own watch; linked worktrees also cover their private gitdir
/// and shared common dir.
#[tauri::command]
pub async fn watch_repo(
    app: tauri::AppHandle,
    state: tauri::State<'_, WatcherState>,
    path: String,
) -> Result<(), CommandError> {
    // Repository discovery plus recursive notify registration touches the
    // filesystem and may be slow on network/large worktrees. WatcherState is
    // Arc-backed so setup can run off the webview thread while callbacks retain
    // the same registry ownership.
    let state = state.inner().clone();
    blocking(move || watcher::watch(&app, &state, &path)).await
}

/// Stop the filesystem watch for `path` (its tab closed).
#[tauri::command]
pub async fn unwatch_repo(
    state: tauri::State<'_, WatcherState>,
    path: String,
) -> Result<(), CommandError> {
    // Recursive watch setup briefly holds the shared registry while installing
    // a replacement. Keep tab-close teardown off the webview thread so it
    // cannot freeze the UI while waiting for that lock.
    let state = state.inner().clone();
    blocking(move || watcher::unwatch(&state, &path)).await
}

File: src/lib/api/git/repo.ts
// Opening a repository and reading its shape: the commit graph, history
// search, branch list, and ancestry queries — plus the filesystem watcher and
// the onboarding path (clone / init / recents). Mirrors `commands/repo.rs`.

import { invoke } from "@/lib/api/invoke";
import { parse } from "@/lib/api/validate";
import { historySearchPageSchema, historySearchResultSchema, repoGraphSchema } from "@/lib/api/schemas";
import { z } from "zod";
import type {
  BranchInfo,
  GitTransportAuthRef,
  HistorySearchPage,
  HistorySearchQuery,
  HistorySearchResult,
  RecentStatus,
  RepoGraph,
  RepoSummary,
} from "./types";

export const repoApi = {
  openRepo: (path: string) => invoke<RepoSummary>("open_repo", { path }),

  commitGraph: async (path: string, limit?: number): Promise<RepoGraph> =>
    parse(repoGraphSchema, await invoke("commit_graph", { path, limit: limit ?? null }), "commit_graph"),

  searchHistory: async (path: string, query: HistorySearchQuery): Promise<HistorySearchPage> =>
    parse(historySearchPageSchema, await invoke("search_history", { path, query }), "search_history"),

  /** HEAD-tree paths (files + directories) containing `filter`, for the
   * advanced search's File-path autosuggest. */
  suggestTreePaths: (path: string, filter: string, limit?: number) =>
    invoke<string[]>("suggest_tree_paths", { path, filter, limit: limit ?? null }),

  listBranches: (path: string) =>
    invoke<BranchInfo[]>("list_branches", { path }),

  /** True when `to` can be fast-forwarded to `from` (from is a descendant of to). */
  canFastForward: (path: string, from: string, to: string) =>
    invoke<boolean>("can_fast_forward", { path, from, to }),

  /** The commits `base..head` would carry, newest first. Graph-only — pair it
   * with [`compareRefs`] for the file/line totals. */
  rangeCommits: async (path: string, base: string, head: string): Promise<HistorySearchResult[]> =>
    parse(
      z.array(historySearchResultSchema),
      await invoke("range_commits", { path, base, head }),
      "range_commits",
    ),

  /** Which of `candidates` `head` descends from, nearest first. Candidates that
   * don't resolve are skipped rather than failing the call. */
  ancestorRefs: async (path: string, head: string, candidates: string[]): Promise<string[]> =>
    parse(
      z.array(z.string()),
      await invoke("ancestor_refs", { path, head, candidates }),
      "ancestor_refs",
    ),

  /** The branch a new pull request from `head` should target by default:
   * gh's `branch.<head>.gh-merge-base` override, else the remote's default
   * branch from `refs/remotes/<remote>/HEAD`. Null when neither is known. */
  defaultBaseBranch: (path: string, head: string) =>
    invoke<string | null>("default_base_branch", { path, head }),

  /** Clone `url` into `dest`, streaming `clone-progress` events while it runs.
   * Resolves with the cloned repo's path; the caller then opens it. Reject with
   * the git failure text (classified UI-side into exists/auth/unreachable). */
  cloneRepo: (url: string, dest: string, auth?: GitTransportAuthRef | null) =>
    invoke<string>("clone_repo", { url, dest, auth: auth ?? null }),

  /** Cancel an in-flight {@link cloneRepo}. Rejects once publication has won
   * the atomic backend race, so callers must keep the success path active. */
  cancelClone: () => invoke<void>("cancel_clone"),

  /** Initialize a new repo at `parent`/`name` on `branch`, optionally seeding a
   * README and a `.gitignore` template. Resolves with the new repo's path. */
  initRepo: (
    parent: string,
    name: string,
    branch: string,
    readme: boolean,
    gitignore: string,
  ) => invoke<string>("init_repo", { parent, name, branch, readme, gitignore }),

  /** Initialize an already-existing, possibly non-empty directory as a repo
   * in place (the missing-repo screen's "Initialize as git repo" recovery
   * action, GL-153) — no README/.gitignore scaffolding. Resolves with the
   * canonical repo path from the post-init open probe. */
  initRepoInPlace: (path: string) => invoke<string>("init_repo_in_place", { path }),

  /** Presence + current branch for each recent repo path (missing-path + branch
   * info for the onboarding "Recent" list). */
  recentsStatus: (paths: string[]) => invoke<RecentStatus[]>("recents_status", { paths }),

  /** Reveal `path` in the OS file manager (Finder/Explorer). */
  revealPath: (path: string) => invoke<void>("reveal_path", { path }),

  /** Start watching `path` (one watch per open tab); the backend emits
   * `repo-changed` events tagged with this path on any change. Linked
   * worktrees also cover their private gitdir and shared common dir. */
  watchRepo: (path: string) => invoke<void>("watch_repo", { path }),

  /** Stop watching `path` (its tab closed). Unknown paths are a no-op. */
  unwatchRepo: (path: string) => invoke<void>("unwatch_repo", { path }),
};
